src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -178,121 +173,95 @@ PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecEle
-
-      if (blocks > 1)
+      blocks = GetAvailableBlocks (allocDuration, spInterval, MAX_NUM_BLOCKS);
+      
+      if (blocks.size() <= 1)
         {
-          while (counter < blocks)
-            {
-              NS_LOG_DEBUG ("Reserve from " << startPeriodicAllocation << " for " << allocDuration);
-
-              endAlloc = startPeriodicAllocation + allocDuration + m_guardTime;
-              UpdateAvailableSlots (startPeriodicAllocation, endAlloc);
-
-              startPeriodicAllocation += spInterval;
-              counter++;
-
-              // update remaining DTI time for consistency
-              m_remainingDtiTime -= (allocDuration + m_guardTime);
-            }
-
-          AddAllocationPeriod (info.GetAllocationID (), info.GetAllocationType (), info.IsPseudoStatic (),
-                               sourceAid, info.GetDestinationAid (),
-                               startFirstAllocation, allocDuration, spInterval, blocks);
-          status.SetSuccess ();
+          // if we cannot guarantee AT LEAST TWO periodic SPs, the request is rejected
+          status.SetFailure ();
+          return status;
         }
-      else
+    }
+    else
+    {
+      blocks = GetAvailableBlocks (allocDuration, 0, 1);
+      if (blocks.size() == 0)
         {
-          // if we cannot guarantee AT LEAST TWO periodic SPs, the request is rejected.
+          // if we cannot guarantee the SP allocation, the request is rejected
           status.SetFailure ();
+          return status;
         }
-
     }
-  else
+    
+  uint32_t startAlloc = blocks[0];
+  uint32_t endAlloc;
+  uint8_t counter = 0;
+  
+  while (counter < blocks.size())
     {
+      NS_LOG_DEBUG ("Reserve from " << startAlloc << " for " << allocDuration);
 
-      uint32_t endAlloc;
-      uint32_t slotDur;
-      for (auto it = m_availableSlots.begin (); it != m_availableSlots.end (); ++it)
-        {
-          slotDur = it->second - it->first;
-
-          if (slotDur > allocDuration)
-            {
-
-              endAlloc = AllocateSingleContiguousBlock (info.GetAllocationID (), info.GetAllocationType (), info.IsPseudoStatic (),
-                                                        sourceAid, info.GetDestinationAid (), it->first, allocDuration);
-              m_remainingDtiTime -= (allocDuration + m_guardTime);
-
-              // The following function modifies m_availableSlots, on which this
-              // loop is iterating. Commonly, it is a bad practice but, as we break 
-              // the loop if we enter this condition, the damage is avoided.
-              UpdateAvailableSlots (it->first, endAlloc);
-              status.SetSuccess ();
-              break;
-            }
-          else
-            {
-              status.SetFailure ();
-            }
-        }
+      endAlloc = startAlloc + allocDuration + m_guardTime;
+      UpdateAvailableSlots (startAlloc, endAlloc);
 
+      startAlloc += spInterval;
+      counter++;
```

Comment:
reuse what you already computed: single point of failure
```suggestion
  uint32_t endAlloc;
  for (i = 0; i < blocks.size(); ++i)
    {
      NS_LOG_DEBUG ("Reserve from " << blocks[i] << " for " << allocDuration);

      endAlloc = blocks[i] + allocDuration + m_guardTime;
      UpdateAvailableSlots (blocks[i], endAlloc);
    }
```

Reply:
true, I didn't think to this possible improvement